ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,858 @@
+not clearly exhibit these semantics. In particular, some records which we
+consider to be events actually lead to the execution of business logic to
+reproduce the state changes they represent, instead of capturing those state
+changes as something that has already happened.
+
+The following is an example of how the BPMN Element Records (which are currently
+all events) can be split into commands and events. You can read this as: when
+processing the `Element_Activating` command, we change the state and write the
+`Element_Activated` event that represents and contains these state changes.
+
+Commands | Events 
+---|---
+Element_Activating | Element_Activated 
+Element_Completing | Element_Completed
+Element_Terminating | Element_Terminated
+Event_Occur (?) | Event_Occurred (?)
+/ | Sequence_Flow_Taken
+
+> Note, that only the `Event_Occur` currently does not exist in Zeebe. It is at
+> this time not yet clear whether we need the `Event_Occur` command and/or the
+> `Event_Occurred` event.
+
+> Also note, that `Sequence_Flow_Taken` is an event without a command. We'll
+> need to find a way to make this fit the strict commands/events separation.
+
+## Processing and Replay
+
+The following diagram shows from left to right a log stream consisting of
+command and event records. Command records can only be processed, while event
+records can only be replayed.
+
+![event-sourcing](ZEP-wf-stream-processing/event-sourcing.png)
+
+> The black arrow indicates that particular events and commands are appended to
+> the log stream as a result of processing a previous command (i.e. it
+> represents the source record position - explained later in more detail). The
+> orange box wrapping some records indicates that these were appended to log
+> stream atomically.
+
+### Activities during Processing
+
+The **commands** are processed only on processing and drive the workflow execution. The processing of a command includes:
+* read the state of the workflow instance and related resources (RocksDB)
+* check if the record can be applied (i.e. concurrent update guard)
+* apply the business logic  
+* write updates to the state
+* write events and (optionally) follow-up commands to continue the processing
+* apply side-effects (e.g. send a response)
+
+The following diagram shows the command processing loop.
+
+![command-processing](ZEP-wf-stream-processing/event-sourcing-processing.png)
+
+### Activities during Replay
+
+The **events** are processed only on replay and rebuild the state. The replay of an event includes:
+* write updates to the state
+
+The following diagram shows the event replay.
+
+![event-replay](ZEP-wf-stream-processing/event-sourcing-replay.png)
+
+In order to rebuild the state with the events, it is necessary that all state changes from the command processing can be applied with the events only.
+So, the command processing may only update the state when it also writes an event that reflects the state changes.
```

Comment:
One idea I'm playing around with in my head is to not make state changes on command processing. Instead prepare events to write and apply their state changes using the events, then write them to the stream.

It may get complicated in situations where state in many different places must be changed and these changes are dependent on each other. For example `CallActivity:ElementActivating` -> applies input mappings and creates a new workflow instance and copies its own variables (dependant on input mappings afaik) to child workflow instance.

Reply:
Hm, I like the concept. We can get around the additional serializing/deserializing by processing the events as is, without reading them back from the log - it still means more processing...but maybe we can just wrap that in one big RocksDB transaction to amortize the cost there? It's a good idea on paper anyway :slightly_smiling_face:

Reply:
Yeah I would avoid reading from the log, since records are appended to the end of the stream. During the processing of the next command all state changes from the previous should be applied already.

Reply:
I agree with Nico's idea. We will see when we start implementing it :balloon: